File: Packages/BioMedLit/Sources/BioMedLit/Transparency/Services/CrossRefService.swift
// BMLibrarian Lite - Biomedical Literature Research Tool
// Copyright (C) 2024-2025 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import Foundation

/// Service for querying the CrossRef API.
///
/// CrossRef provides standardized funder information via the Funder Registry,
/// including funder DOIs that enable reliable industry classification.
///
/// Usage:
/// ```swift
/// let service = CrossRefService(email: "user@example.com")
/// let work = try await service.getWork(doi: "10.1000/example")
/// let funders = service.extractFunders(from: work)
/// ```
public actor CrossRefService {

    // MARK: - Properties

    /// URLSession used for network requests.
    private let session: URLSession

    /// Contact email for CrossRef polite pool identification.
    private let email: String

    /// Timestamp of last request for rate limiting.
    private var lastRequestTime: Date = .distantPast

    // MARK: - Initialization

    /// Initialize the CrossRef service.
    ///
    /// - Parameters:
    ///   - email: Contact email (required by CrossRef for polite pool).
    ///   - session: URLSession to use for requests. If nil, creates a new session
    ///     with default timeout configuration.
    public init(email: String, session: URLSession? = nil) {
        self.email = email

        if let session = session {
            self.session = session
        } else {
            let config = URLSessionConfiguration.default
            config.timeoutIntervalForRequest = BioMedLitConstants.defaultRequestTimeout
            self.session = URLSession(configuration: config)
        }
    }

    // MARK: - API Methods

    /// Get work metadata by DOI.
    ///
    /// Fetches bibliographic metadata for a publication using its DOI, including
    /// funder information, title, journal, and other details.
    ///
    /// - Parameter doi: Digital Object Identifier (e.g., "10.1000/example" or
    ///   "https://doi.org/10.1000/example").
    /// - Returns: Work metadata dictionary from CrossRef API, or nil if not found.
    /// - Throws: CrossRefError on network failure, invalid DOI, or parse error.
    public func getWork(doi: String) async throws -> [String: Any]? {
        // Clean DOI
        let cleanDOI = cleanDOI(doi)

        // Build URL
        let urlString = "\(TransparencyConstants.crossRefBaseURL)/works/\(cleanDOI)"
        guard let url = URL(string: urlString) else {
            throw CrossRefError.invalidDOI(doi)
        }

        // Create request with polite headers
        var request = URLRequest(url: url)
        request.setValue(
            "StudyTransparencyAnalyzer/1.0 (mailto:\(email))",
            forHTTPHeaderField: "User-Agent"
        )

        // Rate limit
        await enforceRateLimit()

        BioMedLitLib.logger?.debug("Fetching CrossRef work for DOI: \(cleanDOI)", category: .network)

        // Execute with retry
        let data = try await RetryHelper.retry(
            config: .networkDefault,
            shouldRetry: RetryHelper.retryOnlyTransient
        ) {
            let (data, response) = try await self.session.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                throw CrossRefError.networkError("Invalid response")
            }

            if BioMedLitConstants.retryableStatusCodes.contains(httpResponse.statusCode) {
                throw CrossRefError.serverError(statusCode: httpResponse.statusCode)
            }

            guard httpResponse.statusCode == BioMedLitConstants.httpStatusOK else {
                if httpResponse.statusCode == BioMedLitConstants.httpStatusNotFound {
                    return nil
                }
                throw CrossRefError.httpError(statusCode: httpResponse.statusCode)
            }

            return data
        }

        guard let responseData = data else {
            BioMedLitLib.logger?.debug("CrossRef returned 404 for DOI: \(cleanDOI)", category: .network)
            return nil
        }

        // Parse JSON
        guard let json = try JSONSerialization.jsonObject(with: responseData) as? [String: Any],
              let message = json["message"] as? [String: Any] else {
            throw CrossRefError.parseError("Invalid JSON structure")
        }

        BioMedLitLib.logger?.info("CrossRef returned work for DOI: \(cleanDOI)", category: .network)

        return message
    }

    // MARK: - Funder Extraction

    /// Extract funder information from CrossRef work.
    ///
    /// Parses the "funder" array from CrossRef work metadata and creates
    /// classified FunderInfo objects using FundingAnalyzer.
    ///
    /// - Parameter work: Work dictionary from CrossRef API.
    /// - Returns: List of FunderInfo objects with industry classifications.
    public func extractFunders(from work: [String: Any]?) -> [FunderInfo] {
        guard let work = work,
              let funders = work["funder"] as? [[String: Any]] else {
            return []
        }

        return FundingAnalyzer.parseCrossRefFunders(funders)
    }

    /// Extract title from CrossRef work.
    ///
    /// - Parameter work: Work dictionary from CrossRef API.
    /// - Returns: Title string if available, nil otherwise.
    public func extractTitle(from work: [String: Any]?) -> String? {
        guard let work = work,
              let titles = work["title"] as? [String],
              let title = titles.first else {
            return nil
        }
        return title
    }

    /// Extract journal name from CrossRef work.
    ///
    /// - Parameter work: Work dictionary from CrossRef API.
    /// - Returns: Journal name if available, nil otherwise.
    public func extractJournal(from work: [String: Any]?) -> String? {
        guard let work = work,
              let containers = work["container-title"] as? [String],
              let journal = containers.first else {
            return nil
        }
        return journal
    }

    /// Extract author names from CrossRef work.
    ///
    /// - Parameter work: Work dictionary from CrossRef API.
    /// - Returns: List of author names in "Family, Given" format.
    public func extractAuthors(from work: [String: Any]?) -> [String] {
        guard let work = work,
              let authors = work["author"] as? [[String: Any]] else {
            return []
        }

        return authors.compactMap { author -> String? in
            let family = author["family"] as? String
            let given = author["given"] as? String

            if let family = family, let given = given {
                return "\(family), \(given)"
            } else if let family = family {
                return family
            }
            return nil
        }
    }

    /// Extract publication date from CrossRef work.
    ///
    /// - Parameter work: Work dictionary from CrossRef API.
    /// - Returns: Publication date if available, nil otherwise.
    public func extractPublicationDate(from work: [String: Any]?) -> Date? {
        guard let work = work else { return nil }

        // Try published-print first, then published-online
        let dateKeys = ["published-print", "published-online", "issued"]

        for key in dateKeys {
            if let dateStruct = work[key] as? [String: Any],
               let dateParts = dateStruct["date-parts"] as? [[Int]],
               let parts = dateParts.first {
                return dateFromParts(parts)
            }
        }

        return nil
    }

    // MARK: - Private Helpers

    /// Clean DOI for URL usage.
    ///
    /// Removes URL prefixes and encodes the DOI for safe use in a URL path.
    ///
    /// - Parameter doi: Raw DOI string.
    /// - Returns: Cleaned and URL-encoded DOI.
    private func cleanDOI(_ doi: String) -> String {
        doi.replacingOccurrences(of: "https://doi.org/", with: "")
           .replacingOccurrences(of: "http://doi.org/", with: "")
           .addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? doi
    }

    /// Enforce rate limiting between requests.
    ///
    /// Ensures minimum interval between requests to comply with CrossRef polite pool.
    private func enforceRateLimit() async {
        let elapsed = Date().timeIntervalSince(lastRequestTime)
        let minInterval = TransparencyConstants.minimumRequestInterval

        if elapsed < minInterval {
            let delay = minInterval - elapsed
            try? await Task.sleep(nanoseconds: UInt64(delay * Double(BioMedLitConstants.nanosecondsPerSecond)))
        }

        lastRequestTime = Date()
    }

    /// Convert date parts array to Date.
    ///
    /// - Parameter parts: Array of [year, month, day] integers (month and day optional).
    /// - Returns: Date if valid parts provided, nil otherwise.
    private func dateFromParts(_ parts: [Int]) -> Date? {
        guard !parts.isEmpty else { return nil }

        var components = DateComponents()
        components.year = parts[0]

        if parts.count > 1 {
            components.month = parts[1]
        } else {
            components.month = 1
        }

        if parts.count > 2 {
            components.day = parts[2]
        } else {
            components.day = 1
        }

        return Calendar.current.date(from: components)
    }
}

// MARK: - Errors

/// Errors that can occur during CrossRef operations.
public enum CrossRefError: LocalizedError, RetryableError, Sendable {
    /// The DOI format is invalid or cannot be used to construct a URL.
    case invalidDOI(String)
    /// A network error occurred (e.g., no connection).
    case networkError(String)
    /// HTTP error response with non-retryable status code.
    case httpError(statusCode: Int)
    /// Server error with retryable status code (5xx, 429).
    case serverError(statusCode: Int)
    /// Failed to parse the API response.
    case parseError(String)

    public var errorDescription: String? {
        switch self {
        case .invalidDOI(let doi):
            return "Invalid DOI: \(doi)"
        case .networkError(let message):
            return "Network error: \(message)"
        case .httpError(let statusCode):
            return "HTTP error: \(statusCode)"
        case .serverError(let statusCode):
            return "Server error (HTTP \(statusCode)). Retrying..."
        case .parseError(let message):
            return "Failed to parse response: \(message)"
        }
    }

    public var isRetryable: Bool {
        switch self {
        case .serverError, .networkError:
            return true
        case .invalidDOI, .httpError, .parseError:
            return false
        }
    }
}

File: android/MedicalFactChecker/app/src/main/java/com/bmlibrarian/factchecker/data/remote/transparency/CrossRefService.kt
/*
 * BMLibrarian Lite - Biomedical Literature Research Tool
 * Copyright (C) 2024-2025 Dr Horst Herb
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.bmlibrarian.factchecker.data.remote.transparency

import android.util.Log
import com.bmlibrarian.factchecker.domain.transparency.FunderInfo
import com.bmlibrarian.factchecker.domain.transparency.FundingAnalyzer
import com.bmlibrarian.factchecker.domain.transparency.JsonCasts
import com.bmlibrarian.factchecker.domain.transparency.TransparencyConstants
import java.io.IOException
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.util.concurrent.TimeUnit
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request

/**
 * Why a CrossRef request failed. Messages are those of the Swift `CrossRefError`.
 *
 * @property isRetryable Whether another attempt could succeed (network and server failures).
 */
sealed class CrossRefException(message: String) : Exception(message) {
    abstract val isRetryable: Boolean

    /** The DOI cannot be turned into a request URL. */
    class InvalidDoi(val doi: String) : CrossRefException("Invalid DOI: $doi") {
        override val isRetryable: Boolean = false
    }

    /** The exchange failed in transport (no connection, timeout, ...). */
    class NetworkError(val detail: String) : CrossRefException("Network error: $detail") {
        override val isRetryable: Boolean = true
    }

    /** A non-retryable HTTP status other than 404. */
    class HttpError(val statusCode: Int) : CrossRefException("HTTP error: $statusCode") {
        override val isRetryable: Boolean = false
    }

    /** A retryable status (429 or 5xx). */
    class ServerError(val statusCode: Int) :
        CrossRefException("CrossRef server error (HTTP $statusCode). Try again later.") {
        override val isRetryable: Boolean = true
    }

    /** The answer was not the JSON shape CrossRef returns. */
    class ParseError(val detail: String) : CrossRefException("Failed to parse response: $detail") {
        override val isRetryable: Boolean = false
    }
}

/**
 * Client for the CrossRef REST API, the source of Funder Registry DOIs that make
 * industry classification reliable.
 *
 * Ported from the Swift `CrossRefService` (BioMedLit): the same URL, polite
 * User-Agent, pacing, retry rule and field parsing. A 404 answers null — the
 * work is absent; every other failure throws a [CrossRefException], so a caller
 * can tell an unreachable CrossRef from a work it does not have.
 *
 * @param email Contact email for the CrossRef polite pool, sent in the User-Agent.
 * @param httpClient HTTP client; defaults to one with Swift's 45 s timeout.
 * @param baseUrl API base URL (overridable for tests).
 * @param retryPolicy How transient failures are retried.
 * @param minimumRequestIntervalMs Minimum gap between two requests.
 */
class CrossRefService(
    private val email: String,
    private val httpClient: OkHttpClient = defaultTransparencyHttpClient(),
    private val baseUrl: String = TransparencyConstants.CROSSREF_BASE_URL,
    private val retryPolicy: TransparencyRetryPolicy = TransparencyRetryPolicy(),
    minimumRequestIntervalMs: Long = TransparencyConstants.MINIMUM_REQUEST_INTERVAL_MS,
) {
    private val pacer = RequestPacer(minimumRequestIntervalMs)

    /**
     * Fetch a work's metadata by DOI.
     *
     * @param doi The DOI, bare ("10.1000/example") or as a doi.org URL.
     * @return The work (CrossRef's `message` object), or null if CrossRef answered 404.
     * @throws CrossRefException on any other failure, after retrying transient ones.
     */
    suspend fun getWork(doi: String): JsonObject? {
        val cleanDoi = cleanDoi(doi)
        val url = "$baseUrl/works/$cleanDoi".toHttpUrlOrNull() ?: throw CrossRefException.InvalidDoi(doi)
        val request = Request.Builder()
            .url(url)
            .header(USER_AGENT_HEADER, "$USER_AGENT_PRODUCT (mailto:$email)")
            .build()

        pacer.awaitTurn()
        Log.d(TAG, "Fetching CrossRef work for DOI: $cleanDoi")

        val body = withTransparencyRetries(retryPolicy, { it is CrossRefException && it.isRetryable }) {
            val response = try {
                executeTransparencyRequest(httpClient, request)
            } catch (e: IOException) {
                throw CrossRefException.NetworkError(e.javaClass.simpleName)
            }
            val status = response.statusCode
            when {
                status in TransparencyHttpConstants.RETRYABLE_STATUS_CODES -> throw CrossRefException.ServerError(status)
                status == TransparencyHttpConstants.HTTP_STATUS_OK -> response.body
                status == TransparencyHttpConstants.HTTP_STATUS_NOT_FOUND -> null
                else -> throw CrossRefException.HttpError(status)
            }
        }

        if (body == null) {
            Log.d(TAG, "CrossRef returned 404 for DOI: $cleanDoi")
            return null
        }

        val message = try {
            JsonCasts.obj(JsonCasts.obj(Json.parseToJsonElement(body))?.get(MESSAGE_KEY))
        } catch (e: SerializationException) {
            null
        } ?: throw CrossRefException.ParseError("Invalid JSON structure")

        Log.i(TAG, "CrossRef returned work for DOI: $cleanDoi")
        return message
    }

    /**
     * Classify the funders of a work's `funder` array.
     *
     * The array is read all-or-nothing, as Swift's `as? [[String: Any]]` is: an
     * entry that is not an object discards every funder.
     *
     * @param work The work, or null.
     * @return The classified funders; empty when there are none.
     */
    fun extractFunders(work: JsonObject?): List<FunderInfo> {
        val funders = JsonCasts.objectList(work?.get("funder")) ?: return emptyList()
        return FundingAnalyzer.parseCrossRefFunders(funders)
    }

    /**
     * The work's title: the first element of its `title` array.
     *
     * @param work The work, or null.
     * @return The title, or null.
     */
    fun extractTitle(work: JsonObject?): String? = JsonCasts.stringList(work?.get("title"))?.firstOrNull()

    /**
     * The journal: the first element of the work's `container-title` array.
     *
     * @param work The work, or null.
     * @return The journal name, or null.
     */
    fun extractJournal(work: JsonObject?): String? =
        JsonCasts.stringList(work?.get("container-title"))?.firstOrNull()

    /**
     * The work's authors as "Family, Given", or "Family" alone when there is no given name.
     *
     * Authors without a family name (e.g. consortia given as `name`) are skipped, as in Swift.
     *
     * @param work The work, or null.
     * @return The author names, in order.
     */
    fun extractAuthors(work: JsonObject?): List<String> {
        val authors = JsonCasts.objectList(work?.get("author")) ?: return emptyList()
        return authors.mapNotNull { author ->
            val family = JsonCasts.string(author["family"]) ?: return@mapNotNull null
            val given = JsonCasts.string(author["given"])
            if (given != null) "$family, $given" else family
        }
    }

    /**
     * The publication date, from the first of `published-print`, `published-online`
     * and `issued` that carries integer `date-parts`.
     *
     * A missing month or day defaults to 1; the date is midnight in the device's
     * time zone, as Swift's `Calendar.current` builds it.
     *
     * @param work The work, or null.
     * @return The date, or null.
     */
    fun extractPublicationDate(work: JsonObject?): Instant? {
        if (work == null) return null
        for (key in DATE_KEYS) {
            val dateStruct = JsonCasts.obj(work[key]) ?: continue
            val parts = JsonCasts.intListList(dateStruct["date-parts"])?.firstOrNull() ?: continue
            return dateFromParts(parts)
        }
        return null
    }

    /** Strip a doi.org URL prefix and percent-encode the DOI for a URL path. */
    private fun cleanDoi(doi: String): String =
        percentEncodePath(doi.replace(HTTPS_DOI_PREFIX, "").replace(HTTP_DOI_PREFIX, ""))

    /** A `[year, month?, day?]` date-parts array as an instant (month and day default to 1). */
    private fun dateFromParts(parts: List<Int>): Instant? {
        val year = parts.firstOrNull() ?: return null
        val month = parts.getOrNull(1) ?: TransparencyConstants.DEFAULT_MONTH
        val day = parts.getOrNull(2) ?: TransparencyConstants.DEFAULT_DAY
        // Lenient like Foundation's Calendar: out-of-range months and days roll over.
        return LocalDate.of(year, 1, 1)
            .plusMonths((month - 1).toLong())
            .plusDays((day - 1).toLong())
            .atStartOfDay(ZoneId.systemDefault())
            .toInstant()
    }

    private companion object {
        const val TAG = "CrossRefService"
        const val USER_AGENT_HEADER = "User-Agent"
        const val USER_AGENT_PRODUCT = "StudyTransparencyAnalyzer/1.0"
        const val MESSAGE_KEY = "message"
        const val HTTPS_DOI_PREFIX = "https://doi.org/"
        const val HTTP_DOI_PREFIX = "http://doi.org/"
        val DATE_KEYS = listOf("published-print", "published-online", "issued")
    }
}

/**
 * An HTTP client with Swift's transparency request timeout, for callers that do
 * not supply the app's own client.
 *
 * @return A new client.
 */
fun defaultTransparencyHttpClient(): OkHttpClient =
    OkHttpClient.Builder()
        .callTimeout(TransparencyHttpConstants.REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS)
        .readTimeout(TransparencyHttpConstants.REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS)
        .build()
